### DISCUSSION TITLE
x:Name on a button isn't getting generated in code behind file.

### DISCUSSION THREAD

**Discussion by gowda2233** [AvaloniaUI/Avalonia#17255] [Q&A] [answered]
### Describe the bug

Following error is reported `ListedDataView.axaml.cs(42,38,42,53): error CS0103: The name 'mySettingButton' does not exist in the current context` when I try to access `mySettingButton` in code behind file.
I have pasted majority of XAML file, the line in question is `<Button x:Name="mySettingButton" IsHitTestVisible="True" ToolTip.Tip="Additional Options">`
Am I doing something wrong here, or is it a bug?

`
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             xmlns:vm="clr-namespace:WhiteRiver.Client.ViewModels"
             x:Class="WhiteRiver.Client.Views.ListedDataView"
             x:DataType="vm:ListedDataViewModel">

	<UserControl.Resources>
		<vm:ColumnMonthValidationConverter x:Key="ColumnMonthValidationConverter"/>
		<vm:ColumnYearValidationConverter x:Key="ColumnYearValidationConverter"/>
		<vm:ColumnDateValidationConverter x:Key="ColumnDateValidationConverter"/>
	</UserControl.Resources>
	<DockPanel>
		<DataGrid x:Name="myDataGrid"
					DockPanel.Dock="Bottom"
					ItemsSource="{Binding ListElements}"
					SelectionMode="Single"
					VerticalScrollBarVisibility="Auto"
					PointerWheelChanged="DataGridPointerWheelChanged"
					VerticalScroll="DataGridVerticalScroll"
					RowHeight="{Binding RowHeight}"
					Margin="0"
					IsReadOnly="False"
					SelectedItem="{Binding ActiveItem}"
					ScrollViewer.HorizontalScrollBarVisibility="Hidden"
					VerticalAlignment="Stretch">
			<DataGrid.KeyBindings>
				<KeyBinding Gesture="Enter" Command="{Binding EnterCommand}" />
			</DataGrid.KeyBindings>
			<DataGrid.ContextMenu>
				<ContextMenu HorizontalAlignment="Stretch" Padding="0">
					<StackPanel  HorizontalAlignment="Stretch">
						<Button Command="{Binding Path=AddToGrabList}" IsHitTestVisible="True" ToolTip.Tip="Add to Grab List" IsVisible="{Binding !GrabMode}">
							<PathIcon Data="{StaticResource add_square_regular}" />
						</Button>
						<Button Command="{Binding Path=EditItem}" IsHitTestVisible="True" ToolTip.Tip="Edit Metadata">
							<PathIcon Data="{StaticResource text_edit_style_regular}" />
						</Button>
						<Button Command="{Binding Path=DownloadImage}" IsHitTestVisible="True" ToolTip.Tip="Download Image">
							<PathIcon Data="{StaticResource arrow_download_regular}" />
						</Button>
						<Button Command="{Binding Path=SoftDeleteImage}" IsHitTestVisible="True" ToolTip.Tip="Remove image from the grablist" IsVisible="{Binding GrabMode}">
							<PathIcon Data="{StaticResource dismiss_regular}" />
						</Button>
						<Button Command="{Binding Path=DeleteImage}" IsHitTestVisible="True" ToolTip.Tip="Delete Image" IsVisible="{Binding !GrabMode}">
							<PathIcon Data="{StaticResource delete_forever_regular}" />
						</Button>
					</StackPanel>
				</ContextMenu>
			</DataGrid.ContextMenu>
			<DataGrid.Columns>
				<DataGridTemplateColumn IsReadOnly="True" Tag="options">
					<DataGridTemplateColumn.CellTemplate>
						<DataTemplate>
							<StackPanel Spacing="8" Margin="0">
								<Button Command="{Binding Path=DownloadImage}" IsHitTestVisible="True" ToolTip.Tip="Download image" IsVisible="{Binding GrabMode}">
									<PathIcon Data="{StaticResource arrow_download_regular}" />
								</Button>
								<Button Command="{Binding Path=AddToGrabList}" IsHitTestVisible="True" ToolTip.Tip="Add to Grab List" IsVisible="{Binding !GrabMode}">
									<PathIcon Data="{StaticResource add_square_regular}" />
								</Button>
								<Button x:Name="mySettingButton" IsHitTestVisible="True" ToolTip.Tip="Additional Options">
									<Button.Flyout>
										<Flyout ShowMode="Transient">
											<StackPanel Width="48" HorizontalAlignment="Center">
												<Button Click="CloseFlyout" Command="{Binding Path=EditItem}" IsHitTestVisible="True" ToolTip.Tip="Edit Metadata">
													<PathIcon Data="{StaticResource text_edit_style_regular}" />
												</Button>
												<Button Click="CloseFlyout" Command="{Binding Path=DownloadImage}" IsHitTestVisible="True" ToolTip.Tip="Download Image" IsVisible="{Binding !GrabMode}">
													<PathIcon Data="{StaticResource arrow_download_regular}" />
												</Button>
												<Button Click="CloseFlyout" Command="{Binding Path=SoftDeleteImage}" IsHitTestVisible="True" ToolTip.Tip="Remove image from the grablist" IsVisible="{Binding GrabMode}">
													<PathIcon Data="{StaticResource dismiss_regular}" />
												</Button>
												<Button Click="CloseFlyout" Command="{Binding Path=DeleteImage}" IsHitTestVisible="True" ToolTip.Tip="Delete Image" IsVisible="{Binding !GrabMode}">
													<PathIcon Data="{StaticResource delete_forever_regular}" />
												</Button>
											</StackPanel>
										</Flyout>
									</Button.Flyout>
									<PathIcon Data="{StaticResource options_regular}" />
								</Button>
							</StackPanel>
						</DataTemplate>
					</DataGridTemplateColumn.CellTemplate>
				</DataGridTemplateColumn>
				<DataGridTemplateColumn>
					<DataGridTemplateColumn.CellTemplate>
						<DataTemplate>
							<Button Command="{Binding Path=PreviewImage}" Padding="0">
								<Button.Styles>
									<Style Selector="Button:pointerover /template/ ContentPresenter">
										<Setter Property="Cursor" Value="Hand" />
									</Style>
								</Button.Styles>
								<Image Source="{Binding Path=Thumbnail}" HorizontalAlignment="Stretch" VerticalAlignment="Stretch"/>
							</Button>
						</DataTemplate>
					</DataGridTemplateColumn.CellTemplate>
				</DataGridTemplateColumn >
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding RRFilter.IsVisible}" CellEditingTemplate="{StaticResource MyRREditTemplate}" HeaderTemplate="{StaticResource MyRRHeaderTemplate}"  CellTemplate="{StaticResource MyRRValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding MonthFilter.IsVisible}" CellEditingTemplate="{StaticResource MyMonthEditTemplate}" HeaderTemplate="{StaticResource MyMonthHeaderTemplate}" CellTemplate="{StaticResource MyMonthValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding DateFilter.IsVisible}" CellEditingTemplate="{StaticResource MyDateEditTemplate}" HeaderTemplate="{StaticResource MyDateHeaderTemplate}" CellTemplate="{StaticResource MyDateValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding YearFilter.IsVisible}" CellEditingTemplate="{StaticResource MyYearEditTemplate}" HeaderTemplate="{StaticResource MyYearHeaderTemplate}" CellTemplate="{StaticResource MyYearValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding IDFilter.IsVisible}" CellEditingTemplate="{StaticResource MyIDEditTemplate}" HeaderTemplate="{StaticResource MyIDHeaderTemplate}" CellTemplate="{StaticResource MyIDValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding LocationFilter.IsVisible}" CellEditingTemplate="{StaticResource MyLocationEditTemplate}" HeaderTemplate="{StaticResource MyLocationHeaderTemplate}" CellTemplate="{StaticResource MyLocationValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding StateFilter.IsVisible}" CellEditingTemplate="{StaticResource MyStateEditTemplate}" HeaderTemplate="{StaticResource MyStateHeaderTemplate}" CellTemplate="{StaticResource MyStateValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding EquipmentFilter.IsVisible}" CellEditingTemplate="{StaticResource MyEquipmentEditTemplate}" HeaderTemplate="{StaticResource MyEquipmentHeaderTemplate}" CellTemplate="{StaticResource MyEquipmentValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding BuilderFilter.IsVisible}" CellEditingTemplate="{StaticResource MyBuilderEditTemplate}" HeaderTemplate="{StaticResource MyBuilderHeaderTemplate}" CellTemplate="{StaticResource MyBuilderValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding ModelFilter.IsVisible}" CellEditingTemplate="{StaticResource MyModelEditTemplate}" HeaderTemplate="{StaticResource MyModelHeaderTemplate}" CellTemplate="{StaticResource MyModelValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding DirFilter.IsVisible}" CellEditingTemplate="{StaticResource MyDirEditTemplate}" HeaderTemplate="{StaticResource MyDirHeaderTemplate}" CellTemplate="{StaticResource MyDirValueTemplate}"/>
				<DataGridTemplateColumn Width="6*"  IsVisible="{Binding PhotoFilter.IsVisible}" CellEditingTemplate="{StaticResource MyPhotoNameEditTemplate}" HeaderTemplate="{StaticResource MyPhotoNameHeaderTemplate}" CellTemplate="{StaticResource MyPhotoNameValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding FileFilter.IsVisible}" HeaderTemplate="{StaticResource MyFileHeaderTemplate}" CellTemplate="{StaticResource MyFileValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding ScanTechFilter.IsVisible}" CellEditingTemplate="{StaticResource MyScanTechEditTemplate}" HeaderTemplate="{StaticResource MyScanTechHeaderTemplate}" CellTemplate="{StaticResource MyScanTechValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding TypeFilter.IsVisible}" CellEditingTemplate="{StaticResource MyTypeEditTemplate}" HeaderTemplate="{StaticResource MyTypeHeaderTemplate}" CellTemplate="{StaticResource MyTypeValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding OwnerFilter.IsVisible}" CellEditingTemplate="{StaticResource MyOwnerEditTemplate}" HeaderTemplate="{StaticResource MyOwnerHeaderTemplate}" CellTemplate="{StaticResource MyOwnerValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding CollectionFilter.IsVisible}" CellEditingTemplate="{StaticResource MyCollectionEditTemplate}" HeaderTemplate="{StaticResource MyCollectionHeaderTemplate}" CellTemplate="{StaticResource MyCollectionValueTemplate}"/>
				<DataGridTemplateColumn Width="4*"  IsVisible="{Binding PublishedFilter.IsVisible}" CellEditingTemplate="{StaticResource MyPublishedEditTemplate}" HeaderTemplate="{StaticResource MyPublishedHeaderTemplate}" CellTemplate="{StaticResource MyPublishedValueTemplate}"/>
				<DataGridTemplateColumn Width="10*" IsVisible="{Binding NotesFilter.IsVisible}" CellEditingTemplate="{StaticResource MyNotesEditTemplate}" HeaderTemplate="{StaticResource MyNotesHeaderTemplate}" CellTemplate="{StaticResource MyNotesValueTemplate}"/>
			</DataGrid.Columns>
		</DataGrid>
	</DockPanel>
</UserControl>
`

### To Reproduce

I am not entirely sure, the error seem to be happening in a particular file, many other files let me declare , x:Name and its field is accessible in the code behind file.

### Expected behavior

mySettingButton file in the code behind file must be accessible 

### Avalonia version

11.0.10

### OS

Windows

**Comment by stevemonaco** [comment, marked as answer]
That `x:Name` is deep within a `DataTemplate`. These are in a different scope and not available in the root's code-behind. Collection controls instantiate many copies of a `DataTemplate` and therefore there would be name collisions.

**Reply by gowda2233** [reply]
Ok, makes much more sense now. Thanks for the response. I will handle the usecase differently.
